core/src/main/java/org/apache/calcite/rel/rules/SortRemoveRedundantRule.java

```diff
@@ -97,13 +110,15 @@ protected SortRemoveRedundantRule(final SortRemoveRedundantRule.Config config) {
   }
 
   private Optional<Double> getSortInputSpecificMaxRowCount(Sort sort) {
-    // If the sort is pure limit, the specific max row count is limit's fetch.
-    if (RelOptUtil.isPureLimit(sort)) {
-      final double limit =
-          sort.fetch instanceof RexLiteral ? RexLiteral.intValue(sort.fetch) : -1D;
-      return Optional.of(limit);
+    if (RelOptUtil.isLimit(sort)) {
+      final double fetch =
+          sort.fetch instanceof RexLiteral ? RexLiteral.intValue(sort.fetch) : 0D;
+      if (fetch == 0D) {
+        return Optional.empty();
+      }
+      // If sort is 'order by x limit n', the target max row count is 1.
+      return RelOptUtil.isOrder(sort) ? Optional.of(1D) : Optional.of(fetch);
     } else if (RelOptUtil.isPureOrder(sort)) {
```

Comment:
what if you place a check first in this method:
* limit 0 (does that even happen?)
* check `isOrder`  ( rows >=1 implied from prev conditional)
this will kinda do the same...

Reply:
Get your point